From 39c1abd0d4e6ea54560d4de38d0d058e3bebefc7 Mon Sep 17 00:00:00 2001
From: David Pacheco <dap@joyent.com>
Date: Wed, 11 Jan 2017 12:21:33 -0800
Subject: [PATCH] MANTA-3080 muskie picker only picks from 200 nodes Reviewed
 by: Bryan Cantrill <bryan@joyent.com> Approved by: Bryan Cantrill
 <bryan@joyent.com> Approved by: Jordan Hendricks
 <jordan.hendricks@joyent.com>

---
 lib/picker.js | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/lib/picker.js b/lib/picker.js
index 5f72e72..6c0bc35 100644
--- a/lib/picker.js
+++ b/lib/picker.js
@@ -82,6 +82,7 @@ var fetch = function fetch_moray(opts, cb) {
     cb = once(cb);
 
     var count = 0;
+    var recs = 0;
     var f = sprintf('(&(percentUsed<=%d)(timestamp>=%d)%s)',
                     opts.utilization || 90,
                     Date.now() - opts.lag,
@@ -103,10 +104,17 @@ var fetch = function fetch_moray(opts, cb) {
         values.push(data.value);
         count = data._count;
         marker = data._id;
+        recs++;
     });
 
     req.once('end', function () {
-        if (values.length < count) {
+        /*
+         * We only fetch "limit" records, but there may be many more storage
+         * nodes than that.  If we saw fewer records than the number that Moray
+         * reported matched our query, that means there are more to fetch, so
+         * we take another lap.
+         */
+        if (recs < count) {
             var next = {
                 lag: opts.lag,
                 limit: opts.limit,
-- 
2.21.0

